plugin/pulpcore/plugin/download/asyncio/group.py

```diff
@@ -0,0 +1,190 @@
+        """
+        if self.group_iterator:
+            # This is not the first call to schedule an iterable
+            self.group_iterator = chain(self.group_iterator, group_iterator)
+            return
+
+        self.group_iterator = group_iterator
+        for i in range(parallel_group_limit):
+            with suppress(StopIteration):
+                self.schedule_group(**next(self.group_iterator))
+
+    def schedule_group(self, id, deferred_artifacts):
+        """
+        Schedules a group of `deferred_artifacts` for downloading, referred to by an `id`.
+
+        Args:
+            id (hashable): This id is handed back to the caller when all files in the group are
+                downloaded.
+            deferred_artifacts (list): A list of :class:`pulpcore.plugin.models.DeferredArtifact`
+                instances that have not been saved to the database.
+        """
+        group = Group(id, deferred_artifacts)
+        self.groups_not_done.append(group)
+        for url in group.urls:
+            if len(self.urls[url]) == 0:
+                # This is the first time we've seen this url so make a downloader
+                downloader_for_url = self.download_factory.build(url)
+                self.downloads_not_done.add(downloader_for_url)
+            self.urls[url].append(group)
+
+    def _find_and_remove_done_group(self):
+        for index, content_unit in enumerate(self.groups_not_done):
+            if content_unit.done:
+                self.groups_not_done.pop(index)
+                return content_unit
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        """
+        Returns:
+            tuple: The first element is the `id` first specified by the user. The second element is
+                a dictionary where the keys are :class:`pulpcore.plugin.models.DeferredArtifact` and
+                the values are dictionaries containing information about the downloaded file. The
+                keys of this dictionary are 'filename', 'size', 'sha1' , 'md5', 'sha224',
+                'sha256', 'sha384', and 'sha512'.
+        """
+        while self.downloads_not_done:
+            done_this_time, self.downloads_not_done = \
+                self.loop.run_until_complete(asyncio.wait(self.downloads_not_done,
+                                                          return_when=asyncio.FIRST_COMPLETED))
+            with suppress(StopIteration):
+                if self.group_iterator:
+                    self.schedule_group(**next(self.group_iterator))
+            for task in done_this_time:
+                download_result = task.result()
+                for group in self.urls[download_result.url]:
+                    group.handle_download_result(download_result)
+                group = self._find_and_remove_done_group()
+                if group:
+                    return group.finished_downloads
+        content_unit = self._find_and_remove_done_group()
```

Comment:
The *content_unit* is a group, right?  Naming *group* would be much clearer.

Reply:
Agreed. Fixed in next push.